src/main/python/ayab/plugins/ayab_plugin/ayab_control.py

```diff
@@ -54,121 +60,136 @@ class AYABControlKnitResult(Enum):
+        self.imgRow, self.color, self.indexToSend, self.sendBlankLine, self.lastLine = self.ayab_control._knit_func(lineNumber)
+        self.bits = self.ayab_control._choose_needles(self.color, self.indexToSend, self.sendBlankLine)
+
+
 
 def even(x):
     return x % 2 == 0
 
 def odd(x):
     return x % 2 == 1
 
+
 class AYABControl(object):
+
     def __init__(self):
         self.__logger = logging.getLogger(type(self).__name__)
-
-        self._API_VERSION = 0x05
-
-        self._current_state = KnittingState.SETUP
-        self._progress = dict(current_row=0,
-                              total_rows=0,
-                              repeats=0,
-                              color="",
-                              hall_l=0,
-                              hall_r=0,
-                              carriage_type="",
-                              carriage_position=0)
+        self.__API_VERSION = 0x05
+        self.__ayabCom = AyabCommunication()
+        self.__current_state = KnittingState.SETUP
+        self.__progress = Progress()
 
     def close(self):
         self.__ayabCom.close_serial()
 
+    def get_row_multiplier(self):
+        return KnittingMode(self.__knitting_mode).row_multiplier(self.__numColors)
+
     def get_progress(self):
-        return self._progress
+        return self.__progress
+
+    def setImage(self, im: Image):
+        self.__image = im
+
+    def setKnittingMode(self, mode: int):
+        self.__knitting_mode = mode
+
+    def setNumColors(self, n: int):
+        self.__numColors = n
+
+    def setStartLine(self, n: int):
+        self.__startLine = n
+
+    def setInfRepeat(self, x: bool):
+        self.__infRepeat = x
 
     def __get_knit_func(self):
         '''Select function that decides which line of data to send according to the machine type and number of colors'''
-        if self._knitting_mode == KnittingMode.SINGLEBED.value and self._numColors == 2:
-            self._knit_func = self._singlebed_2col
-        elif self._knitting_mode == KnittingMode.CLASSIC_RIBBER_1.value and self._numColors == 2:
-            self._knit_func = self._doublebed_2col
-        elif self._knitting_mode == KnittingMode.CLASSIC_RIBBER_1.value and self._numColors > 2:
-            self._knit_func = self._doublebed_multicol
-        elif self._knitting_mode == KnittingMode.MIDDLECOLORSTWICE_RIBBER.value and self._numColors >= 2:
-            self._knit_func = self._middlecoltwice
-        elif self._knitting_mode == KnittingMode.HEARTOFPLUTO_RIBBER.value and self._numColors >= 2:
-            self._knit_func = self._heartofpluto
-        elif self._knitting_mode == KnittingMode.CIRCULAR_RIBBER.value: # not restricted to 2 colors
-            self._knit_func = self._circular_ribber
-        else:
-            self.__logger.error("Fallthrough error in __get_knit_func: invalid knitting options")
-            return False # knit function not found
+        if not self.__knitting_mode.good_ncolors(self.__numColors):
```

Comment:
self.__knitting_mode needs a cast to KnittingMode().

Reply:
Done